HDDS-7309. Enable by default GRPC between S3G and OM

xBis7 opened the pull request:
## What changes were proposed in this pull request?

After discussions and testing, submitting this patch to enable GRPC server by default between S3G and OM, to improve performance.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-7309

## How was this patch tested?

This patch was tested manually.

kerneltime commented:
cc @duongkame

michelsumbul commented:
@kerneltime As we said at the apacheCon, we will look at having this number when we have a bit more time, however this should not be a prerequisite to enable that feature by default. We already demonstrate that it increase the number of ops the S3G can handle and the user can still disable it if they want. Enabling it by default will allow to have more people to try it and it to be sure it doesnt break in the future with new pr

duongkame commented:
> e by default. We already demonstrate that it increase the number of ops the S3G can handle and the user can still disable it if they want. Enabling it by default will allow to have more people to try it and it to be sure it doesnt break in the future with new pr

+1. I also think that the full gRpc scalability test is not required to enable the protocol on the server-side by default. 

That may be needed when we indeed start using gRpc as the default protocol on the client sides, like S3G or OFS.

duongkame commented on hadoop-hdds/common/src/main/resources/ozone-default.xml:
```diff
@@ -2506,6 +2506,14 @@
   </property>
+  <property>
+    <name>ozone.om.transport.class</name>
+    <value>org.apache.hadoop.ozone.om.protocolPB.GrpcOmTransportFactory</value>
```
So, do we need to make [the actual change](https://github.com/apache/ozone/blob/master/hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/OMConfigKeys.java#L341-L341) to make `GrpcOmTransportFactory` the default value for `ozone.om.transport.class`.

xBis7 replied:
No, `ozone-default.xml` supercedes the `OMConfigKeys` default value. The tests get the transport class from `OMConfigKeys`, so this way hadoop config will continue getting used there.

kerneltime commented:
> > e by default. We already demonstrate that it increase the number of ops the S3G can handle and the user can still disable it if they want. Enabling it by default will allow to have more people to try it and it to be sure it doesnt break in the future with new pr
> 
> +1. I also think that the full gRpc scalability test is not required to enable the protocol on the server-side by default.
> 
> That may be needed when we indeed start using gRpc as the default protocol on the client sides, like S3G or OFS.

Makes sense. I think we should rename the Jira to indicate that this just starts the GRPC server on the OM and does not switch the client in S3G to start using it.

adoroszlai commented:
> this just starts the GRPC server on the OM and does not switch the client in S3G to start using it.

Maybe I'm missing something, but it seems to me this patch **does switch** S3 Gateway to use gRPC for talking to OM by default.

`OmTransportFactory#createFactory` creates the actual factory as an instance of the (first) class specified in the service descriptor, which is `GrpcOmTransportFactory` for S3 Gateway:

https://github.com/apache/ozone/blob/46e58a6350e86f7573fe45678e24c5dab96df45a/hadoop-ozone/s3gateway/src/main/resources/META-INF/services/org.apache.hadoop.ozone.om.protocolPB.OmTransportFactory#L15

if the value configured for `ozone.om.transport.class` does not match the value of `OMConfigKeys#OZONE_OM_TRANSPORT_CLASS_DEFAULT`, which was and still is:

https://github.com/apache/ozone/blob/46e58a6350e86f7573fe45678e24c5dab96df45a/hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/OMConfigKeys.java#L341-L343

Prior to this patch default value of `ozone.om.transport.class` and `OMConfigKeys#OZONE_OM_TRANSPORT_CLASS_DEFAULT` did match, so S3 Gateway used the default transport despite its service descriptor.  Since the value in `ozone-default.xml` is being changed, this condition will now be true by default:

https://github.com/apache/ozone/blob/46e58a6350e86f7573fe45678e24c5dab96df45a/hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OmTransportFactory.java#L48-L54

So `GrpcOmTransportFactory` will be created by default for S3 Gateway, i.e. it will switch to the gRPC client.

From `ozone` compose environment with default configs:

```
s3g_1       | 2022-10-14 19:32:35,747 [qtp661119548-86] INFO protocolPB.OmTransportFactory: ZZZ created custom OmTransportFactory: org.apache.hadoop.ozone.om.protocolPB.GrpcOmTransportFactory@77d1c0db
s3g_1       | 2022-10-14 19:32:35,920 [qtp661119548-86] INFO protocolPB.GrpcOmTransport: GrpcOmTransport: started
```

The first log message is one I added in `OmTransportFactory`.

Also, enabling it only in OM, but not in S3 Gateway, would not have the following benefit:

> Enabling it by default will allow to have more people to try it

xBis7 commented:
@adoroszlai is right. We are enabling gRPC for S3G [here](https://github.com/xBis7/ozone/blob/HDDS-7309/hadoop-hdds/common/src/main/resources/ozone-default.xml#L2511).

kerneltime commented:
> > this just starts the GRPC server on the OM and does not switch the client in S3G to start using it.
> 
> Maybe I'm missing something, but it seems to me this patch **does switch** S3 Gateway to use gRPC for talking to OM by default.
> 
> `OmTransportFactory#createFactory` creates the actual factory as an instance of the (first) class specified in the service descriptor, which is `GrpcOmTransportFactory` for S3 Gateway:
> 
> https://github.com/apache/ozone/blob/46e58a6350e86f7573fe45678e24c5dab96df45a/hadoop-ozone/s3gateway/src/main/resources/META-INF/services/org.apache.hadoop.ozone.om.protocolPB.OmTransportFactory#L15
> 
> if the value configured for `ozone.om.transport.class` does not match the value of `OMConfigKeys#OZONE_OM_TRANSPORT_CLASS_DEFAULT`, which was and still is:
> 
> https://github.com/apache/ozone/blob/46e58a6350e86f7573fe45678e24c5dab96df45a/hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/OMConfigKeys.java#L341-L343
> 
> Prior to this patch default value of `ozone.om.transport.class` and `OMConfigKeys#OZONE_OM_TRANSPORT_CLASS_DEFAULT` did match, so S3 Gateway used the default transport despite its service descriptor. Since the value in `ozone-default.xml` is being changed, this condition will now be true by default:
> 
> https://github.com/apache/ozone/blob/46e58a6350e86f7573fe45678e24c5dab96df45a/hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OmTransportFactory.java#L48-L54
> 
> So `GrpcOmTransportFactory` will be created by default for S3 Gateway, i.e. it will switch to the gRPC client.
> 
> From `ozone` compose environment with default configs:
> 
> ```
> s3g_1       | 2022-10-14 19:32:35,747 [qtp661119548-86] INFO protocolPB.OmTransportFactory: ZZZ created custom OmTransportFactory: org.apache.hadoop.ozone.om.protocolPB.GrpcOmTransportFactory@77d1c0db
> s3g_1       | 2022-10-14 19:32:35,920 [qtp661119548-86] INFO protocolPB.GrpcOmTransport: GrpcOmTransport: started
> ```
> 
> The first log message is one I added in `OmTransportFactory`.
> 
> Also, enabling it only in OM, but not in S3 Gateway, would not have the following benefit:
> 
> > Enabling it by default will allow to have more people to try it

Thank you @adoroszlai.

@xBis7 would be possible to get peak OM performance numbers, I would like to validate that the peak OM performance stays the same or is better with GRPC. I understand that each client is seeing better performance but it would be prudent to check the peak performance on OM as well.

smengcl commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/TestOzoneConfigurationFields.java:
```diff
@@ -66,6 +66,7 @@ public void initializeMemberVariables() {
     xmlPrefixToSkipCompare.add("ipc.client.rpc-timeout.ms");
     xmlPropsToSkipCompare.add("ozone.om.leader.election.minimum.timeout" +
         ".duration"); // Deprecated config
+    xmlPropsToSkipCompare.add("ozone.om.transport.class");
```
You probably want to delete this line:

https://github.com/apache/ozone/blob/a66e43e2838f087229cb4b0fac1e1dc969ab80dd/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/TestOzoneConfigurationFields.java#L103

rather than adding this to skip set?

smengcl replied:
Try deleting this line:

https://github.com/apache/ozone/blob/a66e43e2838f087229cb4b0fac1e1dc969ab80dd/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/TestOzoneConfigurationFields.java#L127

rather than adding it to the skip set?

smengcl commented on hadoop-ozone/integration-test/src/test/resources/ozone-site.xml:
```diff
@@ -26,6 +26,16 @@
     <value>org.apache.hadoop.hdds.fs.MockSpaceUsageCheckFactory$None</value>
   </property>
 
+  <property>
+    <name>ozone.om.transport.class</name>
+    <value>org.apache.hadoop.ozone.om.protocolPB.Hadoop3OmTransportFactory</value>
+  </property>
+
+  <property>
+    <name>ozone.om.s3.grpc.server_enabled</name>
+    <value>false</value>
+  </property>
```
I presume this is to avoid breaking some of the integration tests?

If so, which integration tests are broken once we enable this by default? And how well is the new default tested so far?

xBis7 replied:
@smengcl thanks for looking at this PR. There are too many integration tests failing with GRPC enabled. I launched a workflow to go over the failures. Please take a look [here](https://github.com/xBis7/ozone/actions/runs/3361441746).

> And how well is the new default tested so far?

The performance numbers have been checked by @michelsumbul.

smengcl replied:
I see **most** are just port binding failures. Possibly a docker compose config issue. If the port number has been changed with the enablement of gRPC the docker image might need to open those new ports as well. Try opening up port 8981 in the corresponding `docker-compose.yaml` files?

```
Error:  org.apache.hadoop.fs.ozone.TestOzoneFsHAURLs  Time elapsed: 7.445 s  <<< ERROR!
java.io.IOException: Failed to bind to address 0.0.0.0/0.0.0.0:8981
	at io.grpc.netty.NettyServer.start(NettyServer.java:328)
	at io.grpc.internal.ServerImpl.start(ServerImpl.java:183)
	at io.grpc.internal.ServerImpl.start(ServerImpl.java:92)
	at org.apache.hadoop.ozone.om.GrpcOzoneManagerServer.start(GrpcOzoneManagerServer.java:123)
```
https://github.com/xBis7/ozone/actions/runs/3361441746/jobs/5571845363

```
Error:  Errors: 
Error:    Failed to bind to address 0.0.0.0/0.0.0.0:8981
Error:    Failed to bind to address 0.0.0.0/0.0.0.0:8981
Error:    Failed to bind to address 0.0.0.0/0.0.0.0:8981
Error:    TestAddRemoveOzoneManager.testBootstrap:173->setupCluster:94 » IO Failed to bi...
Error:    TestAddRemoveOzoneManager.testDecommission:347->setupCluster:94 » IO Failed to...
Error:    TestAddRemoveOzoneManager.testForceBootstrap:283->setupCluster:94 » IO Failed ...
Error:    TestOMRatisSnapshots.init:110 » IO Failed to bind to address 0.0.0.0/0.0.0.0:8...
Error:    TestOMRatisSnapshots.init:110 » IO Failed to bind to address 0.0.0.0/0.0.0.0:8...
...
```
https://github.com/xBis7/ozone/actions/runs/3361441746/jobs/5571845635

adoroszlai replied:
@smengcl the failing ones are integration tests, no docker involved.  I think fixing them requires using random port in mini cluster for the grpc server in OM.

smengcl replied:
@adoroszlai You are right. Multiple OMs are trying to bind to the same port, causing this issue.

xBis7 replied:
Mini cluster tries to configure Grpc server ports on multiple OMs with the same port. The Grpc server is not used by any of the integration tests. We disable it in the `ozone-site` because it's not needed, there are no integration tests for the S3G.

dave-gantenbein replied:
Monotonically increasing port numbers, not random, pls. 

#noflakytests

captainzmc commented:
Hi @adoroszlai @smengcl  , In dev email communication, this PR is required by Ozone 1.3. Currently the release of Ozone-1.3 is blocked on this PR. Could you help continue to review this PR? cc @neils-dev

adoroszlai commented:
@captainzmc, there is no need for further review of this PR, it is good for `master`, can be merged.  Backporting it to 1.3 is a different question.

I don't see dev communication that this is blocking the release.  I see a [request to include it in the release if possible](https://lists.apache.org/thread/3701bl0rmc13hckokt38kbvcyw1h7jvz).

I'm not in favor of making such a change so late in the 1.3 release cycle.  Enabling OM gRPC is a matter of simple config change, anyone can do it manually.  I think it makes sense to first ship it disabled by default, let users enable on an as-needed basis and provide feedback, then enable in the next release based on that.

captainzmc commented:
> I'm not in favor of making such a change so late in the 1.3 release cycle. Enabling OM gRPC is a matter of simple config change, anyone can do it manually.

@adoroszlai Agree with you. In fact, users can change the configuration to decide whether to use grpc. I will remove this issue from the block list in 1.3. Then we can start preparing for 1.3.0-rc0.

neils-dev commented:
Thanks @captainzmc for following up for this PR on the 1.3.0-rc branch and @adoroszlai for comments on that.  I did request that this PR be included in the 1.3.0-rc but the comment from @adoroszlai is great, with it merged into the master as the default s3g config and allowing the 1.3.0-rc preparations without it,

> @captainzmc, there is no need for further review of this PR, it is good for master, can be merged (based on the assumption that @smengcl's only concern was integration test coverage).
> ...remove this issue from the block list in 1.3. Then we can start preparing for 1.3.0-rc0.

DaveTeng0 commented:
Hello guys~ looks like this PR is ready to be merged. Maybe we could go ahead to merge?

kerneltime commented:
@DaveTeng0 There is one bit of information we are waiting for, we need to measure the peak OM performance for Hadoop RPC vs GRPC and then this can be merged if there is no regression.

kerneltime commented:
@xBis7 we did some initial tests and OM peak performance is about the same between the 2.
Can you rebase this change?

xBis7 commented:
@adoroszlai Thanks for updating the PR and pointing out the failure. I'll take a look.

xBis7 commented:
We added this file `/hadoop-ozone/integration-test/src/test/resources/META-INF/services/org.apache.hadoop.ozone.om.protocolPB.OmTransportFactory` under `integration-test` which was forcing tests to use Hadoop RPC despite their configurations. The test that was failing, needed to use gRPC and had been configured to do so but this file was overriding the setting. 

I've removed it and run a workflow on a dummy branch successfully.
https://github.com/xBis7/ozone/actions/runs/4544882681

xBis7 commented:
@kerneltime mentioned that there is no performance gain on the OM side. What about the clients? If there is no performance gain, maybe we should reconsider making this the default option.

adoroszlai commented:
Thanks @xBis7 for the patch, @duongkame, @kerneltime, @smengcl for the review.
